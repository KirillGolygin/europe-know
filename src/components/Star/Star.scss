.star {
    padding-top: 13px;
    cursor: pointer;

    .star-icon {
        &:hover {
            fill: black;
        }
    }
}

.filled {
    fill: black
}