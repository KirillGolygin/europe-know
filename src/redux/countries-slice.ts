import { createAction, createSlice } from '@reduxjs/toolkit';
import { put, call } from 'redux-saga/effects';

import type { PayloadAction } from '@reduxjs/toolkit';
import type { RootState } from './store';

import { getAllCountries, getCountryInfo } from '../api';

import type { ICountry, IPickedCountry } from '../interfaces/country';
import { AxiosResponse } from 'axios';

interface CountriesState {
  countries: ICountry[];
  favourits: ICountry[];
  filteredCountries: ICountry[];
  pickedCountry: IPickedCountry | null;
  loading: boolean;
  error: string | null;
}

const initialState: CountriesState = {
  countries: [],
  favourits: [],
  pickedCountry: null,
  filteredCountries: [],
  loading: false,
  error: null
};

export function* getCountriesSaga() {
  try {
    yield put(getCountriesPending());
    const response: AxiosResponse<unknown, unknown> = yield call(getAllCountries);

    const payload: ICountry[] = yield response.data;
    yield put(getCountriesSuccess(payload));
    yield put(updateFavorites());
  } catch (error) {
    yield put(getCountriesRejected(`Произошла ошибка: ${error}`));
  }
}

export function* getCountryDetailsSaga(action: PayloadAction<string>) {
  const response: AxiosResponse<IPickedCountry[], unknown> = yield call(
    getCountryInfo,
    action.payload
  );

  const data: IPickedCountry = yield response.data[0];

  const prepearedData: IPickedCountry = {
    ...data,
    currencies: Object.keys(data.currencies),
    languages: Object.values(data.languages),
    population: Number(data.population).toLocaleString('ru')
  };

  yield put(pickCountry(prepearedData));
  yield put(updateFavorites());
}

export const CountriesSlice = createSlice({
  name: 'countries',
  initialState,
  reducers: {
    getCountriesPending: (state) => {
      state.loading = true;
      state.error = null;
    },
    getCountriesRejected: (state, action: PayloadAction<string>) => {
      state.loading = false;
      state.error = action.payload;
    },
    getCountriesSuccess: (state, action: PayloadAction<ICountry[]>) => {
      state.loading = false;
      state.error = null;
      state.filteredCountries = action.payload;
      state.countries = action.payload;
      state.countries.forEach((country) => (country.favourite = false));
    },
    sortCountries: (state, action: PayloadAction<string>) => {
      state.filteredCountries = state.countries.filter((country) => {
        if (!action.payload) {
          return country.name.common.toLowerCase().includes('');
        } else {
          return country.name.common.toLowerCase().includes(action.payload.toLowerCase());
        }
      });
    },
    changeFavourites: (state, action: PayloadAction<string>) => {
      const countryToChange = state.filteredCountries.find(
        (country) => country.name.common === action.payload
      );

      if (!countryToChange) return;

      if (countryToChange.favourite === true) {
        state.favourits = state.favourits.filter(
          (country) => country.name.common !== countryToChange?.name.common
        );
        if (state.pickedCountry) {
          state.pickedCountry.favourite = false;
        }
        state.filteredCountries.forEach((country) => {
          if (country.name.common === countryToChange.name.common) {
            country.favourite = false;
          }
        });
      } else {
        state.filteredCountries.forEach((country) => {
          if (country.name.common === countryToChange.name.common) {
            country.favourite = true;
          }
        });
        if (state.pickedCountry) {
          state.pickedCountry.favourite = true;
        }
        countryToChange.favourite = true;
        state.favourits.push(countryToChange);
      }
    },
    updateFavorites: (state) => {
      state.favourits.forEach((fav) => {
        if (fav.name.common === state.pickedCountry?.name.common) {
          state.pickedCountry.favourite = true;
        }
        state.filteredCountries.forEach((country) => {
          if (country.name.common === fav.name.common) {
            country.favourite = true;
            state.favourits = state.favourits.filter(
              (old) => old.name.common !== country.name.common
            );
            state.favourits.push(country);
          }
        });
      });
    },
    clearFavourites: (state) => {
      state.favourits = initialState.favourits;
      state.filteredCountries.forEach((country) => (country.favourite = false));
    },
    pickCountry: (state, action: PayloadAction<IPickedCountry>) => {
      state.pickedCountry = action.payload;
    }
  }
});

export const GET_COUNTRIES = 'countries/getCountries';
export const getCountries = createAction(GET_COUNTRIES);

export const GET_COUNTRY_DETAILS = 'countries/getCountryDetails';
export const getCountryDetails = createAction<string>(GET_COUNTRY_DETAILS);

export const {
  getCountriesSuccess,
  getCountriesPending,
  getCountriesRejected,
  sortCountries,
  changeFavourites,
  updateFavorites,
  clearFavourites,
  pickCountry
} = CountriesSlice.actions;

export const selectCountries = (state: RootState) => state.countries.countries;
export const selectFilteredCountries = (state: RootState) => state.countries.filteredCountries;
export const selectFavourites = (state: RootState) => state.countries.favourits;
export const selectPickedCountry = (state: RootState) => state.countries.pickedCountry;
export const selectLoading = (state: RootState) => state.countries.loading;
export const selectError = (state: RootState) => state.countries.error;

export default CountriesSlice.reducer;
