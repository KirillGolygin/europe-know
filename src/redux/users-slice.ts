import { IUser } from "./../interfaces/user";
import { createAction, createSlice } from "@reduxjs/toolkit";
import { put, call, select } from "redux-saga/effects";

import type { PayloadAction } from "@reduxjs/toolkit";
import type { RootState } from "./store";

import { getUsers, regUser } from "../api";

interface UsersState {
  currentUser: IUser | null;
  siginFormData: IUser | null;
  regFormData: IUser | null;
  signinErrorMessage: string | null;
  isAlreadyRegistered: boolean;
  registerSuccessed: boolean;
  error: string | null;
}

const initialState: UsersState = {
  currentUser: null,
  siginFormData: null,
  regFormData: null,
  signinErrorMessage: null,
  error: null,
  registerSuccessed: false,
  isAlreadyRegistered: false,
};

export function* signinUserSaga(): any {
  const { users } = yield select();
  yield put(loginUserLoading());
  try {
    const response = yield call(getUsers);
    const { result } = yield response.data;

    const correctUser = result.find(
      (user: IUser) =>
        user.login === users.siginFormData.login &&
        user.password === users.siginFormData.password
    );
    yield put(loginUserSuccess(correctUser));
  } catch (error) {
    yield put(loginUseRejected(`Произошла ошибка: ${error}`));
  }
}

export function* regUserSaga(): any {
  const { users } = yield select();
  const response = yield call(getUsers);
  const { result } = yield response.data;

  const registeredUser = result.find(
    (user: IUser) => user.login === users.regFormData.login
  );

  if (registeredUser) {
    yield put(toggleIsAlreadyRegistered(true));
  } else {
    yield call(regUser, users.regFormData);
    yield put(toggleRegisterStatus(true));
  }
}

export const usersSlice = createSlice({
  name: "users",
  initialState,
  reducers: {
    loginUserLoading: (state) => {
      state.signinErrorMessage = null;
    },
    loginUserSuccess: (state, action: PayloadAction<IUser>) => {
      if (action.payload) {
        state.currentUser = action.payload;
      } else {
        state.signinErrorMessage = "Данный пользователь не зарегестрирован";
      }
    },
    loginUseRejected: (state, action: PayloadAction<string>) => {
      state.error = action.payload;
    },
    saveSigninFormData: (state, action: PayloadAction<IUser>) => {
      state.siginFormData = action.payload;
    },
    saveRegFormData: (state, action: PayloadAction<IUser>) => {
      state.regFormData = action.payload;
    },
    logoutUser: (state) => {
      state.currentUser = null;
    },
    toggleRegisterStatus: (state, action: PayloadAction<boolean>) => {
      state.registerSuccessed = action.payload;
    },
    toggleIsAlreadyRegistered: (state, action: PayloadAction<boolean>) => {
      state.isAlreadyRegistered = action.payload;
    },
  },
});

export const SIGNIN_USER = "users/signinUser";
export const signinUser = createAction(SIGNIN_USER);

export const REG_USER = "users/regUser";
export const registerUser = createAction(REG_USER);

export const {
  loginUserLoading,
  loginUserSuccess,
  loginUseRejected,
  saveSigninFormData,
  logoutUser,
  toggleRegisterStatus,
  toggleIsAlreadyRegistered,
  saveRegFormData,
} = usersSlice.actions;

export const selectFilteredCountries = (state: RootState) =>
  state.countries.filteredCountries;

export const selectCurrentUser = (state: RootState) => state.users.currentUser;
export const selectSigninError = (state: RootState) =>
  state.users.signinErrorMessage;
export const selectError = (state: RootState) => state.users.error;
export const selectRegisterSuccessed = (state: RootState) =>
  state.users.registerSuccessed;
export const selectIsAlreadyRegistered = (state: RootState) =>
  state.users.isAlreadyRegistered;

export default usersSlice.reducer;
