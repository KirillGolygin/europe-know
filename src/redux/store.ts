import { configureStore } from "@reduxjs/toolkit";
import createSagaMiddleware from "@redux-saga/core";
import { takeEvery } from "redux-saga/effects";

import countriesReducer from "./countries-slice";
import { GET_COUNTRIES, getCountriesSaga } from "./countries-slice";

import popupsReducer from "./popups-slice";

const sagaMiddleware = createSagaMiddleware();

function* sagas() {
  yield takeEvery(GET_COUNTRIES, getCountriesSaga);
}

export const store = configureStore({
  devTools: true,
  reducer: {
    countries: countriesReducer,
    popups: popupsReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      thunk: false,
    }).concat(sagaMiddleware),
});

sagaMiddleware.run(sagas);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
