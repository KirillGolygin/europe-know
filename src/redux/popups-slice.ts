import { PayloadAction, createSlice } from '@reduxjs/toolkit';
import { RootState } from './store';

interface PopupsState {
  popUp: {
    type: 'signin' | 'register';
    isOpen: boolean;
  };
}

const initialState: PopupsState = {
  popUp: {
    type: 'register',
    isOpen: false
  }
};

export const PopupsSlice = createSlice({
  name: 'popups',
  initialState,
  reducers: {
    openPopup: (state, action: PayloadAction<'signin' | 'register'>) => {
      state.popUp.isOpen = true;
      state.popUp.type = action.payload;
    },

    closePopup: (state) => {
      state.popUp.isOpen = false;
    }
  }
});

export const { openPopup, closePopup } = PopupsSlice.actions;

export const selectPopup = (state: RootState) => state.popups.popUp;

export default PopupsSlice.reducer;
