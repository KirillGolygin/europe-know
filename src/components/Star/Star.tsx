/// <reference types="vite-plugin-svgr/client" />

import { useAppSelector, useAppDispatch } from "../../redux/hooks/redux-hooks";

import { selectCurrentUser } from "../../redux/users-slice";
import { openPopup } from "../../redux/popups-slice";
import { changeFavourites } from "../../redux/countries-slice";

import StarIcon from "../../assets/svg/star.svg?react";

import cn from "classnames";
import "./Star.scss";

interface StarProps {
  countryName: string;
  favourite: boolean;
}

const Star = ({ countryName, favourite }: StarProps) => {
  const dispatch = useAppDispatch();
  const changeFavouriteStatus = (name: string) =>
    dispatch(changeFavourites(name));

  const currentUser = useAppSelector(selectCurrentUser);
  return (
    <div
      className="star"
      onClick={
        !currentUser
          ? () => dispatch(openPopup("signin"))
          : () => changeFavouriteStatus(countryName)
      }
    >
      <StarIcon className={cn("star-icon", { ["filled"]: favourite })} />
    </div>
  );
};

export default Star;
