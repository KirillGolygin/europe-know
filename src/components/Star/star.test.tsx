import { expect, it, vi } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";

import * as reduxHooks from "../../redux/hooks/redux-hooks";
import * as popupActions from "../../redux/popups-slice";
import * as CuntriesActions from "../../redux/countries-slice";
import Star from "./Star";

vi.mock("../../redux/hooks/redux-hooks");

const mockedDispatch = vi.spyOn(reduxHooks, "useAppDispatch");
const mockedUseSelector = vi.spyOn(reduxHooks, "useAppSelector");

it("renders correctly when favourite = false", () => {
  mockedDispatch.mockReturnValue(vi.fn());

  const component = render(<Star countryName="Russia" favourite={false} />);
  expect(component).toMatchSnapshot();
});

it("renders correctly when favourite = true", () => {
  mockedDispatch.mockReturnValue(vi.fn());

  const component = render(<Star countryName="Russia" favourite={true} />);
  expect(component).toMatchSnapshot();
});

it("call open popup if not signed in", () => {
  const dispatch = vi.fn();
  mockedDispatch.mockReturnValue(dispatch);
  mockedUseSelector.mockReturnValue(null);

  const mockedOpenPopup = vi.spyOn(popupActions, "openPopup");

  render(<Star countryName="Russia" favourite={false} />);

  const btn = screen.getByRole("button");
  fireEvent.click(btn);
  expect(dispatch).toBeCalledTimes(1);
  expect(mockedOpenPopup).toBeCalled();
});

it("call changeFavourites if signed in", () => {
  const dispatch = vi.fn();
  mockedDispatch.mockReturnValue(dispatch);
  mockedUseSelector.mockReturnValue(true);

  const mockedChangeFavurites = vi.spyOn(CuntriesActions, "changeFavourites");

  render(<Star countryName="Russia" favourite={false} />);

  const btn = screen.getByRole("button");
  fireEvent.click(btn);
  expect(dispatch).toBeCalledTimes(1);
  expect(mockedChangeFavurites).toBeCalled();
});
