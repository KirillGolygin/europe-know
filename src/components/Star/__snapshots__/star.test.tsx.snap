// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly when favourite = false 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="star"
      >
        <div
          class="star-icon"
          data-testid="star-icon"
        >
          <svg
            enable-background="new 0 0 512 512"
            height="25px"
            id="Layer_1"
            viewBox="0 0 512 512"
            width="25px"
            x="0px"
            xml:space="preserve"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
            y="0px"
          >
            <polygon
              points=" 259.216,29.942 330.27,173.919 489.16,197.007 374.185,309.08 401.33,467.31 259.216,392.612 117.104,467.31 144.25,309.08  29.274,197.007 188.165,173.919 "
              stroke="black"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="10"
              stroke-width="37.6152"
            />
          </svg>
        </div>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="star"
    >
      <div
        class="star-icon"
        data-testid="star-icon"
      >
        <svg
          enable-background="new 0 0 512 512"
          height="25px"
          id="Layer_1"
          viewBox="0 0 512 512"
          width="25px"
          x="0px"
          xml:space="preserve"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
          y="0px"
        >
          <polygon
            points=" 259.216,29.942 330.27,173.919 489.16,197.007 374.185,309.08 401.33,467.31 259.216,392.612 117.104,467.31 144.25,309.08  29.274,197.007 188.165,173.919 "
            stroke="black"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-miterlimit="10"
            stroke-width="37.6152"
          />
        </svg>
      </div>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`renders correctly when favourite = true 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="star"
      >
        <div
          class="star-icon"
          data-testid="star-icon"
        >
          <svg
            enable-background="new 0 0 512 512"
            height="25px"
            id="Layer_1"
            viewBox="0 0 512 512"
            width="25px"
            x="0px"
            xml:space="preserve"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
            y="0px"
          >
            <polygon
              points=" 259.216,29.942 330.27,173.919 489.16,197.007 374.185,309.08 401.33,467.31 259.216,392.612 117.104,467.31 144.25,309.08  29.274,197.007 188.165,173.919 "
              stroke="black"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="10"
              stroke-width="37.6152"
            />
          </svg>
        </div>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="star"
    >
      <div
        class="star-icon"
        data-testid="star-icon"
      >
        <svg
          enable-background="new 0 0 512 512"
          height="25px"
          id="Layer_1"
          viewBox="0 0 512 512"
          width="25px"
          x="0px"
          xml:space="preserve"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
          y="0px"
        >
          <polygon
            points=" 259.216,29.942 330.27,173.919 489.16,197.007 374.185,309.08 401.33,467.31 259.216,392.612 117.104,467.31 144.25,309.08  29.274,197.007 188.165,173.919 "
            stroke="black"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-miterlimit="10"
            stroke-width="37.6152"
          />
        </svg>
      </div>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
